example_allocators/freelist_allocator/freelist_allocator.c

```diff
@@ -0,0 +1,166 @@
+#include "freelist_allocator.h"
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+
+char *small_freelist = NULL;
+
+char *medium_freelist = NULL;
+
+char *large_freelist = NULL;
+
+void initialize(unsigned int size_in_pages)
+{
+	/* request memory for our allocation buffer
+	 * NB mmap min bounds for capability is 1 page (4K)
+	 */
+	size_t bytes_to_allocate = size_in_pages * BYTES_IN_PAGE;
+	char *res =
+		mmap(NULL, bytes_to_allocate, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+
+	if (res == MAP_FAILED)
+	{
+		perror("error in initial mem allocation");
+		exit(-1);
+	}
+
+	// put in linked list pointers and
+	// stick into the large freelist
+	// give this space to the large freelist ...
+	large_freelist = insert_linked_list_pointers(LARGE, bytes_to_allocate, res, large_freelist);
+	return;
+}
+
+char *insert_linked_list_pointers(size_t cell_size, size_t limit, char *start, char *freelist)
+{
+	char *curr = start;
+	char *next = curr + cell_size;
+	char *max = start + limit;
+	// ensure next ptr will fit into cell
+	assert(sizeof(void *) <= cell_size);
+
+	while (next < max)
+	{
+		((char **)curr)[0] = next;
+		curr = next;
+		next = curr + cell_size;
+	}
+	// at the end, concatenate this newly formed
+	// list with existing freelist
+	((char **)curr)[0] = freelist;
+
+	return start;
+}
+
+char *alloc(size_t bytes)
+{
+
+	size_t size;
+	char *freelist_to_use;
+	char *larger_freelist;
+	char *ret = NULL; // ptr to return
+
+	// work out which freelist to use
+	if (bytes <= SMALL)
+	{
+		size = SMALL;
+		freelist_to_use = small_freelist;
+	}
+	else if (bytes <= MEDIUM)
+	{
+		size = MEDIUM;
+		freelist_to_use = medium_freelist;
+	}
+	else
+	{
+		size = LARGE;
+		freelist_to_use = large_freelist;
+	}
+
+	if (freelist_to_use == NULL)
+	{
+		// fixup freelist (if no available mem there)
+		char *new_space;
+		switch (size)
+		{
+		case SMALL:
+			new_space = alloc(MEDIUM);
+			if (new_space != NULL)
+			{
+				small_freelist =
+					insert_linked_list_pointers(SMALL, MEDIUM, new_space, small_freelist);
+				freelist_to_use = small_freelist;
+				// now we have replenished space...
+			}
+			break;
+		case MEDIUM:
+			new_space = alloc(LARGE);
+			if (new_space != NULL)
+			{
+				medium_freelist =
+					insert_linked_list_pointers(MEDIUM, LARGE, new_space, medium_freelist);
+				freelist_to_use = medium_freelist;
+			}
+			break;
+		default:
+			// stuck! no more mem!
```

Comment:
Should we `err` at this point or return `NULL` or ... ?

Reply:
it does return NULL in this case, due to the initial value of ret at the start of the function. I will make this clear in a comment